File: LeetCode/binaryTree/222.count-complete-tree-nodes.py
#
# @lc app=leetcode id=222 lang=python3
#
# [222] Count Complete Tree Nodes
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0

        leftHeight = rightHeight = 0
        leftPointer = rightPointer = root

        while (leftPointer is not None):
            leftHeight+=1
            leftPointer = leftPointer.left
       
        while (rightPointer is not None):
            rightHeight+=1
            rightPointer = rightPointer.right

        if (leftHeight == rightHeight):
            return 2**leftHeight-1
        else:
            return 1 + self.countNodes(root.left) + self.countNodes(root.right)

        
# @lc code=end
# O(logn*logn)
# https://labuladong.github.io/algo/2/21/48/

# counting #nodes in a general binary tree
    # def countNodes(root: Optional[TreeNode]) -> int:
    #     if root is None:
    #         return 0
    #     return 1+countNodes(root.left)+countNodes(root.right)
    # O(n)

# count #nodes in a general perfect binary tree:
    # def countNodes(root: Optional[TreeNode]) -> int:
    #     height = 0
    #     while (root is not None):
    #         height+=1
    #         root = root.right
    #     return 2^height-1
    # O(logn)

# Complete binary tree is the combination of the two algorithm, using counting sub-perfect-binary-tree to optimize the complexity

File: LeetCode_Go/BinaryTree/222.count-complete-tree-nodes.go
/*
 * @lc app=leetcode id=222 lang=golang
 *
 * [222] Count Complete Tree Nodes
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// Given the root of a complete binary tree, return the number of the nodes in the tree.

// Notice that no matter how the leaves align in the last level, either The left subtree or the right subtree must be a perfect binary tree, and the other one is a complete/perfect binary tree. For a perfect binary tree we can solve the problem with formula (2^n - 1) in O(1). Then we can recursivly solve the other subtree.

// T(n) = 2log(n) + T(n/2) = ... = log(n)*log(n)

func countNodes(root *TreeNode) int {
	if root == nil {
		return 0
	}
	var leftCount int
	var rightCount int
	// check if the left subtree is a perfect binary tree
	left, right := root.Left, root.Left
	leftHeight, rightHeight := 0, 0
	for left != nil {
		leftHeight++
		left = left.Left
	}
	for right != nil {
		rightHeight++
		right = right.Right
	}
	if leftHeight == rightHeight { // left subtree is a perfect binary tree
		leftCount = 1<<leftHeight - 1
		rightCount = countNodes(root.Right)
	} else { // right subtree is a perfect binary tree of one level less than the left subtree (happens to be rightHeight)
		leftCount = countNodes(root.Left)
		rightCount = 1<<rightHeight - 1
	}
	return leftCount + rightCount + 1
}

// @lc code=end

